alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -623,3 +644,139 @@ instance
+evaluateTransactionExecutionUnits tx utxo ei sysS (CostModel costModel) =
+  Map.fromList $ map findAndCount neededPlutusScripts
+  where
+    note :: e -> Maybe a -> Either e a
+    note _ (Just x) = Right x
+    note e Nothing = Left e
+
+    txb = getField @"body" tx
+    ws = getField @"wits" tx
+    rs = unRedeemers $ getField @"txrdmrs" ws
+    dats = unTxDats $ getField @"txdats" ws
+    scripts = getField @"txscripts" ws
+    sNeeded = scriptsNeeded utxo tx
+    neededPlutusScripts = do
+      (sp, sh) <- sNeeded
+      msb <- case Map.lookup sh scripts of
+        Nothing -> pure Nothing
+        Just (TimelockScript _) -> []
+        Just (PlutusScript bytes) -> pure $ Just bytes
+      pure (sp, sh, msb)
+
+    txinfo = txInfo ei sysS utxo tx
+
+    exBudgetToExUnits :: P.ExBudget -> Maybe ExUnits
+    exBudgetToExUnits (P.ExBudget (P.ExCPU cpu) (P.ExMemory memory)) =
+      ExUnits <$> convertCost cpu <*> convertCost memory
+
+    convertCost :: P.CostingInteger -> Maybe Word64
+    convertCost ci =
+      if (i <= toInteger (maxBound :: Word64)) && (i >= 0)
+        then Just $ fromIntegral i
+        else Nothing
+      where
+        i = toInteger ci
+
+    findAndCount ::
+      ( Alonzo.ScriptPurpose (Crypto era),
+        ScriptHash (Crypto era),
+        Maybe ShortByteString
+      ) ->
+      ( ScriptHash (Crypto era),
+        Either (ScriptFailure (Crypto era)) (RdmrPtr, ExUnits)
+      )
+    findAndCount (sp, sh, mScriptBytes) = (,) sh $ do
+      r <-
+        note (InvalidScriptPurpose sp) $
+          strictMaybeToMaybe $ Alonzo.rdptr @era txb sp
+      (rdmr, _) <- note (MissingRedeemer r) $ Map.lookup r rs
+      script <- note (MissingScript sh) $ mScriptBytes
+      args <- case sp of
+        (Alonzo.Spending txin) -> do
+          txOut <- note (UnknownTxIn txin) $ Map.lookup txin (unUTxO utxo)
+          let TxOut _ _ mdh = txOut
+          dh <- note (InvalidTxIn txin) $ strictMaybeToMaybe mdh
+          dat <- note (MissingDatum dh) $ Map.lookup dh dats
+          pure [dat, rdmr, valContext txinfo sp]
+        _ -> pure [rdmr, valContext txinfo sp]
+      let pArgs = map getPlutusData args
+
+      exUnits <- case snd $ P.evaluateScriptCounting P.Quiet costModel script pArgs of
+        Left e -> Left $ ValidationFailed e
+        Right exBudget -> note (IncompatibleBudget exBudget) $ exBudgetToExUnits exBudget
+
+      pure (r, exUnits)
+
+evaluateTransactionBalance ::
+  forall era.
+  ( Era era,
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    HasField "mint" (Core.TxBody era) (Core.Value era),
+    HasField "wdrls" (Core.TxBody era) (Wdrl (Crypto era)),
+    HasField "_keyDeposit" (Core.PParams era) Coin,
+    HasField "_poolDeposit" (Core.PParams era) Coin
+  ) =>
+  Core.PParams era ->
+  UTxO era ->
+  Map (KeyHash 'StakePool (Crypto era)) (PoolParams (Crypto era)) ->
```

Comment:
I can make this `Set (KeyHash 'StakePool (Crypto era))` if that is better. I wanted to re-use `produced`, but I wasn't sure if it was a bad idea to change `produced` to take a set, since then it is computing the domain of this mapping for every transaction, and all it needs to do is use it for a key lookup.

Reply:
this is now `Set (KeyHash 'StakePool (Crypto era)) -> Bool`, ie `Map.notMember` curried in the second parameter